contract_comm/currency/currency.go

```diff
@@ -0,0 +1,270 @@
+	// (val * n1 * d2) / (d1 * n2)
+	numerator := new(big.Int).Mul(val, new(big.Int).Mul(exchangeRateFrom.Numerator, exchangeRateTo.Denominator))
+	denominator := new(big.Int).Mul(exchangeRateFrom.Denominator, exchangeRateTo.Numerator)
+	return new(big.Int).Div(numerator, denominator), nil
+}
+
+func Cmp(val1 *big.Int, currency1 *common.Address, val2 *big.Int, currency2 *common.Address) int {
+	if currency1 == currency2 {
+		return val1.Cmp(val2)
+	}
+
+	exchangeRate1, err1 := getExchangeRate(currency1)
+	exchangeRate2, err2 := getExchangeRate(currency2)
+
+	if err1 != nil || err2 != nil {
+		currency1Output := "nil"
+		if currency1 != nil {
+			currency1Output = currency1.Hex()
+		}
+		currency2Output := "nil"
+		if currency2 != nil {
+			currency2Output = currency2.Hex()
+		}
+		log.Warn("Error in retrieving exchange rate.  Will do comparison of two values without exchange rate conversion.", "currency1", currency1Output, "err1", err1, "currency2", currency2Output, "err2", err2)
+		return val1.Cmp(val2)
+	}
+
+	// Below code block is basically evaluating this comparison:
+	// val1 * exchangeRate1.Numerator/exchangeRate1.Denominator < val2 * exchangeRate2.Numerator/exchangeRate2.Denominator
+	// It will transform that comparison to this, to remove having to deal with fractional values.
+	// val1 * exchangeRate1.Numerator * exchangeRate2.Denominator < val2 * exchangeRate2.Numerator * exchangeRate1.Denominator
+	leftSide := new(big.Int).Mul(val1, new(big.Int).Mul(exchangeRate1.Numerator, exchangeRate2.Denominator))
+	rightSide := new(big.Int).Mul(val2, new(big.Int).Mul(exchangeRate2.Numerator, exchangeRate1.Denominator))
+	return leftSide.Cmp(rightSide)
+}
+
+func getExchangeRate(currencyAddress *common.Address) (*exchangeRate, error) {
+	var (
+		returnArray [2]*big.Int
+		leftoverGas uint64
+	)
+
+	if currencyAddress == nil {
+		return &exchangeRate{cgExchangeRateNum, cgExchangeRateDen}, nil
+	} else {
+		if leftoverGas, err := contract_comm.MakeStaticCall(params.SortedOraclesRegistryId, medianRateFuncABI, "medianRate", []interface{}{currencyAddress}, &returnArray, 20000, nil, nil); err != nil {
+			if err == params.ErrSmartContractNotDeployed {
+				log.Warn("Registry address lookup failed", "err", err)
+				return &exchangeRate{big.NewInt(1), big.NewInt(1)}, err
+			} else {
+				log.Error("medianRate invocation error", "gasCurrencyAddress", currencyAddress.Hex(), "leftoverGas", leftoverGas, "err", err)
+				return &exchangeRate{big.NewInt(1), big.NewInt(1)}, err
+			}
+		}
+	}
+	log.Trace("medianRate invocation success", "gasCurrencyAddress", currencyAddress, "returnArray", returnArray, "leftoverGas", leftoverGas)
+	return &exchangeRate{returnArray[0], returnArray[1]}, nil
+}
+
+// This function will retrieve the balance of an ERC20 token.
+//
+func GetBalanceOf(accountOwner common.Address, contractAddress common.Address, evm *vm.EVM, gas uint64) (result *big.Int, gasUsed uint64, err error) {
```

Comment:
This is the only function within any of the contract_comm subpackages that accepts a vm.evm object.  All others will accept a state and header object.

For consistency, I suggest that that this function also takes in a state and header object.

Reply:
Great point - done.